### PR TITLE
🌐make sure the source repo are get only once during restore and build

### PR DESCRIPTION
**Pull request by OsmondJiang**
that's a pre-work for #4383

### CODE REVIEW

File: src/docfx/build/Build.cs [hunk 1]
```diff
@@ -266,5 +286,29 @@ private static IReadOnlyList<Document> GetTableOfContentsScope(Docset docset, Ta
 
             return result;
         }
+
+        private class SourceRepo
```
Comment (yufeih) [inline src/docfx/build/Build.cs]: we already have a `Repository` class, can that be reused?

Comment (OsmondJiang) [reply src/docfx/build/Build.cs]: `Repository` need have "path"

File: src/docfx/build/Build.cs [hunk 2]
```diff
@@ -18,32 +18,44 @@ public static async Task Run(string docsetPath, CommandLineOptions options, Repo
...
-        private static async Task Run(string docsetPath, Repository repository, string locale, CommandLineOptions options, Report report, DependencyLockModel dependencyLock, RestoreMap restoreMap)
+        private static async Task Run(
+            string docsetPath,
+            Repository repository,
+            string locale,
+            CommandLineOptions options,
+            Report report,
+            DependencyLockModel dependencyLock,
+            RestoreMap restoreMap,
+            Repository sourceRepo = null)
```
Comment (yufeih) [inline src/docfx/build/Build.cs]: consider name it `fallbackRepo` or `locFallbackRepo`. `sourceRepo` could also mean source/target repo for a pull request.

File: src/docfx/build/Build.cs [hunk 3]
```diff
@@ -197,41 +209,50 @@ void ValidateBookmarks()
             }
         }
 
-        private static DependencyLockModel LoadBuildDependencyLock(string docset, string locale, Repository repository, CommandLineOptions commandLineOptions)
+        private static (DependencyLockModel dependencyLock, (string remote, string branch) sourceRepoInfo) LoadBuildDependencyLock(
```
Comment (yufeih) [inline src/docfx/build/Build.cs]:
```suggestion
        private static (DependencyLockModel dependencyLock, string remote, string branch) LoadBuildDependencyLock(
```

File: src/docfx/localization/LocalizationUtility.cs [hunk 4]
```diff
@@ -237,6 +182,42 @@ public static bool TryRemoveLocale(string name, out string nameWithoutLocale, ou
             return false;
         }
 
+        internal static bool TryGetSourceRepository(string remote, string branch, out string sourceRemote, out string sourceBranch, out string locale)
```
Comment (yufeih) [inline src/docfx/localization/LocalizationUtility.cs]: nit: Still make it public to make diff smaller?

```suggestion
        public static bool TryGetSourceRepository(string remote, string branch, out string sourceRemote, out string sourceBranch, out string locale)
```

File: src/docfx/build/Build.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/build/Build.cs]: good idea

File: src/docfx/localization/LocalizationUtility.cs (hunk 4 above)
Comment (OsmondJiang) [reply src/docfx/localization/LocalizationUtility.cs]: or let me try to make it private

**Pull request merged**